src/Symfony/Component/DependencyInjection/ReverseContainer.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException;
+
+/**
+ * Turns public and "container.reversible" services back to their ids.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ReverseContainer
+{
+    private $serviceContainer;
+    private $reversibleLocator;
+    private $getServiceId;
+
+    public function __construct(Container $serviceContainer, ContainerInterface $reversibleLocator, string $tagName = 'container.reversible')
+    {
+        $this->serviceContainer = $serviceContainer;
+        $this->reversibleLocator = $reversibleLocator;
+        $this->getServiceId = \Closure::bind(function ($service) {
+            return array_search($service, $this->services, true) ?: array_search($service, $this->privates, true);
+        }, $serviceContainer, Container::class);
+    }
+
+    /**
+     * @param object $service
+     *
+     * @throws ServiceNotFoundException When the service is not reversible and $throw is true
+     */
+    public function getId($service, bool $throw = true): ?string
+    {
+        if ($this->serviceContainer === $service) {
+            return 'service_container';
+        }
+
+        if (null === $id = ($this->getServiceId)($service)) {
+            if (!$throw) {
+                return null;
+            }
+
+            $id = \get_class($service);
+            throw new ServiceNotFoundException($id, null, null, [], sprintf('Service of type "%s" not found: please ensure the instance is registered as a service and either make it public or tag it as "%s".', $id, $this->tagName));
+        }
+
+        if ($this->serviceContainer->has($id) || $this->reversibleLocator->has($id)) {
+            return $id;
+        }
+
+        if (!$throw) {
+            return null;
+        }
+
+        throw new ServiceNotFoundException($id, null, null, [], sprintf('The "%s" service is private and cannot be accessed by reference. You should either make it public, or tag it as "%s".', $id, $this->tagName));
+    }
+
+    /**
+     * @return object
+     *
+     * @throws ServiceNotFoundException When the service is not reversible
+     */
+    public function getService(string $id)
```

Comment:
Why not let people use a normal service locator for this?

Reply:
Because it's not a normal service locator: you can get here only the ids you can fetch with the other method before. It would make sense to provide a standard locator interface here - that's not what this service is for.

Reply:
isn't it the same as having a service locator for services tagged with `container.reversible` that you can easily do manually?

Reply:
those, + public services

Reply:
So why does it need the method then?

Reply:
Because an instance of this class is a single consistent scope of services you can reverse.
Providing only object->id mapping here, and relying on another object to do the reverse would mean the two objects are bound by some external convention that cannot be enforced by any contracts. Having both methods in this class is what provide the needed guarantee, from the abstraction pov.